### PR TITLE
235 connection to cognito form

### PR DESCRIPTION
**Pull request by rnudb**

### CODE REVIEW

File: interview/views.py [hunk 1]
```diff
@@ -1464,3 +1464,55 @@ def interviews_list(request):
...
+        process.save()
+
+        if "Document" in data:
+            for document in data["Document"]:
+                r = requests.get(document["File"], stream=True)
+                file_tmp = NamedTemporaryFile(delete=True, suffix=".pdf")
+                file_tmp.write(r.content)
+                file_tmp.flush()
+                Document.objects.create(document_type="CV", content=File(file_tmp), candidate=candidate)
+                print("Created document")
```
Comment (bport) [inline interview/views.py:1513]: Delete print

File: interview/views.py [hunk 2]
```diff
@@ -1464,3 +1464,55 @@ def interviews_list(request):
...
+        "subsidiary": subsidiary_id,
+    }
+    if "Offer_Value" in data:
+        process_data["offer"] = data["Offer_Value"]
+
+    process_data["other_informations"] = ""
+    if "Motivation" in data:
+        process_data["other_informations"] = data["Motivation"]
+    if "Availability" in data:
+        process_data["other_informations"] += " - Disponible dès: " + data["Availability"]
```
Comment (bport) [inline interview/views.py:1497]: We have a dedicated field for that

File: interview/views.py [hunk 3]
```diff
@@ -1464,3 +1464,55 @@ def interviews_list(request):
...
+    if process_form.is_valid() and candidate_form.is_valid():
+        candidate = candidate_form.save()
+        process = process_form.save(commit=False)
+        process.candidate = candidate
+        process.save()
+
+        if "Document" in data:
+            for document in data["Document"]:
+                r = requests.get(document["File"], stream=True)
+                file_tmp = NamedTemporaryFile(delete=True, suffix=".pdf")
```
Comment (bport) [inline interview/views.py:1509]: I guess we can have something else than pdf, use file name to get extension

File: pyoupyou/settings/dev.py [hunk 4]
```diff
@@ -52,3 +52,5 @@
 
 SEEKUBE_SOURCE_ID = 5
 SECRET_ANON_SALT = "dev-salt"
+
+FORM_WEB_HOOK_PREFIX = "wefstcaticnbhmnhbkyqbbaqwxlxgbta"
```
Comment (bport) [inline pyoupyou/settings/dev.py:56]: replace with "dev" easiest for test locally and we don't want security on our local machine

File: interview/views.py [hunk 5]
```diff
@@ -1464,3 +1464,55 @@ def interviews_list(request):
...
+    data = json.loads(request.body)
+
+    candidate_data = {
+        "name": data["Name"],
+        "email": data["Email"],
+    }
+    if "Phone" in data:
+        candidate_data["phone"] = data["Phone"]
+    if "Linkedin" in data:
+        candidate_data["linkedin_url"] = data["Linkedin"]
```
Comment (bport) [inline interview/views.py:1482]: This field don't exist ?

File: pyoupyou/urls.py [hunk 6]
```diff
@@ -29,6 +30,11 @@
     url(r"^processes/offer/(?P<offer_id>\d+)$", views.processes_for_offer, name="process-list-offer"),
     url(r"^interviews/$", views.interviews_list, name="interviews-list"),
     url(r"^candidate/$", views.new_candidate, name="candidate-new"),
+    url(
+        r"^webhook/" + settings.FORM_WEB_HOOK_PREFIX + r"/(?P<subsidiary_id>\d+)/(?P<source_id>\d+)$",
+        csrf_exempt(views.process_from_cognito_form),
```
Comment (bport) [inline pyoupyou/urls.py:35]: set csrf_exempt as decorator on the function directly

File: interview/tests.py [hunk 7]
```diff
@@ -330,3 +334,119 @@ def test_anonymize_candidate(self):
...
+        self.subsidiary = SubsidiaryFactory()
+        self.url = "/webhook/{prefix}/{sub_id}/{source_id}".format(
+            source_id=self.source.id, sub_id=self.subsidiary.id, prefix=settings.FORM_WEB_HOOK_PREFIX
+        )
+
+        self.consultant = ConsultantFactory(company=self.subsidiary)
+        self.user = self.consultant.user
+
+        # create 6 offer as offer id in request is 5
+        for i in range(6):
```
Comment (bport) [inline interview/tests.py:351]: Tips change json so we don't need to create 6 :-)

**Review by bport — CHANGES_REQUESTED** [review]

File: interview/views.py [hunk 8]
```diff
@@ -1465,3 +1465,56 @@ def interviews_list(request):
     }
 
     return render(request, "interview/list_interviews.html", context)
+
+
+@csrf_exempt
+@require_http_methods(["POST"])
+def process_from_cognito_form(request, source_id, subsidiary_id):
+    # TODO: rebase on 226 to have linkedin field in candidate
```
Comment (bport) [inline interview/views.py:1473]: I guess it's done :-)

File: interview/tests.py [hunk 9]
```diff
@@ -330,3 +334,69 @@ def test_anonymize_candidate(self):
...
+        self.source = SourcesFactory()
+        self.subsidiary = SubsidiaryFactory()
+        self.url = "/webhook/{prefix}/{sub_id}/{source_id}".format(
+            source_id=self.source.id, sub_id=self.subsidiary.id, prefix=settings.FORM_WEB_HOOK_PREFIX
+        )
+
+        self.consultant = ConsultantFactory(company=self.subsidiary)
+        self.user = self.consultant.user
+
+        # create 6 offer as offer id in request is 5
```
Comment (bport) [inline interview/tests.py:350]: can remove comment :-)

File: interview/tests.py [hunk 10]
```diff
@@ -330,3 +334,69 @@ def test_anonymize_candidate(self):
...
+        self.client.force_login(self.user)
+        data = {
+            "Form": {
+                "Id": "1",
+                "InternalName": "Formulaire",
+                "Name": "Formulaire de candidature",
+            },
+            "$version": 8,
+            "$etag": "W/\"datetime'2022-11-04T13%3A56%3A25.06651Z'\"",
+            "Name": self.candidate.name,
```
Comment (bport) [inline interview/tests.py:365]: Don't use factory there, hard code a name, email and phone

**Review by bport — CHANGES_REQUESTED** [review]

File: interview/views.py [hunk 11]
```diff
@@ -1465,3 +1465,57 @@ def interviews_list(request):
...
+
+        if process_form.is_valid() and candidate_form.is_valid():
+            candidate = candidate_form.save()
+            process = process_form.save(commit=False)
+            process.candidate = candidate
+            process.save()
+
+            if "Document" in data:
+                for document in data["Document"]:
+                    r = requests.get(document["File"], stream=True)
```
Comment (ArnaudChirat) [inline interview/views.py:1509]: You need to close the file

File: interview/tests.py [hunk 12]
```diff
@@ -330,3 +334,76 @@ def test_anonymize_candidate(self):
         self.assertEqual(other_candidate2.anonymized_email(), email_hash.hexdigest())
         previous_candidate_anoymized = other_candidate2.find_duplicates()
         self.assertEqual(1, previous_candidate_anoymized.count())
+
+
+class ImportCognitoFormTestCase(TestCase):
```
Comment (ArnaudChirat) [inline interview/tests.py:339]: You need a test with documents

File: interview/serializers.py [hunk 13]
```diff
@@ -0,0 +1,56 @@
...
+            sources=Sources.objects.get(id=validated_data["sources"]),
+            subsidiary=Subsidiary.objects.get(id=validated_data["subsidiary"]),
+            offer=Offer.objects.filter(id=validated_data["Offer_Value"]).first(),
+            contract_start_date=validated_data["Availability"],
+            other_informations=validated_data["Motivation"] or "",
+        )
+
+        documents = validated_data.pop("Document", [])
+        for document in documents:
+            r = requests.get(document["File"], stream=True)
```
Comment (ArnaudChirat) [inline interview/serializers.py:35]: how do you get the request ?

File: interview/views.py [hunk 14]
```diff
@@ -1465,3 +1467,24 @@ def interviews_list(request):
...
+    data.update(
+        {
+            "sources": source_id,
+            "subsidiary": subsidiary_id,
+        }
+    )
+
+    serializer = CognitoWebHookSerializer(data=data)
+
+    serializer.is_valid(raise_exception=True)
```
Comment (ArnaudChirat) [inline interview/views.py:1486]: return bad request if not valid (even if cognito doesnt care, it is usefull for debug)

**Review by ArnaudChirat — APPROVED** [review]

**Pull request merged**